src/System.Memory/src/System/MemoryExtensions.Portable.cs

```diff
@@ -12,6 +12,34 @@ namespace System
     /// </summary>
     public static partial class MemoryExtensions
     {
+        /// <summary>
+        /// Determines whether the end of the <paramref name="span"/> matches the specified <paramref name="value"/> when compared using the specified <paramref name="comparison"/> option.
+        /// </summary>
+        /// <param name="span">The source span.</param>
+        /// <param name="value">The sequence to compare to the end of the source span.</param>
+        /// <param name="comparison">One of the enumeration values that determines how the <paramref name="span"/> and <paramref name="value"/> are compared.</param>
+        public static bool EndsWith(this ReadOnlySpan<char> span, ReadOnlySpan<char> value, StringComparison comparison)
+        {
+            string sourceString = span.ToString();
+            string valueString = value.ToString();
```

Comment:
Ouch; this is really unfortunate.  We should consider only exposing these for core.

If we keep these, we should at least do a reasonable implementation for Ordinal and OrdinalIgnoreCase, and presumably we could do so for Invariant as well?  I agree the CurrentCulture ones will be problematic.

Reply:
Invariant would be as the current culture (if the string is not fully ASCII). 

I think there is opportunity to optimize for the ordinal and ordinal ignore casing but I think we already saying the portable implementation would be slower than the core one and whatever we do will not be able to match the core performance (i guess).

not exposing these would be an issue for anyone want to write a library that work on net core and down level too. I prefer keeping it and document it as it is slower than the core one and we can try optimize here as much we can.

---
In reply to: [166251624](https://github.com/dotnet/corefx/pull/26880#discussion_r166251624) [](ancestors = 166251624)

Reply:
I can file an issue to optimize this method for the ordinal cases. We do have the generic StartsWith method that take a T, which does not take SringComparison that can be used for one of the two scenarios.

FYI
>     // Those need access to globalization APIs. We'll also expose them from
    // the .NET Framework OOB (slow span). They will try to extract the string
    // from the underlying span (because slow span stores it) -- or -- allocate
    // a new string. This avoids bifurcating the API surface.

Reply:
> They will try to extract the string from the underlying span (because slow span stores it)

Can this change do this then?

Reply:
I haven't been able to figure out how to do that. The underlying data is stored as a Pinnable<char>, and the globalization APIs work on strings (not char*).

Reply:
It looks like we're using Unsafe.As to cast the object to a Pinnable.  Can you just do a typecheck on it and cast it back? e.g. `object o = _pinnable; if (o is string s) { ... }`?

Reply:
If that is doable, I can definitely update our ToString implementation and take advantage of that as well.

Reply:
Is there a way to do something similar for the FastSpan ToString which has a ByReference field?

Reply:
No, there is not a reasonable way to do that for fast span.